Google.Api.Generator/ProtoUtils/ResourceDetails.cs

```diff
@@ -25,148 +26,296 @@ internal class ResourceDetails
+                    var requireIdentifier = !((fieldDesc.IsRepeated && fieldDesc.Name.ToLowerInvariant() == "names") ||
+                        (!fieldDesc.IsRepeated && fieldDesc.Name.ToLowerInvariant() == "name"));
+                    var requireAs = requireIdentifier || isWildcard;
+                    var requirePlural = fieldDesc.IsRepeated;
+                    var name = requireIdentifier ? UnderlyingPropertyName : "";
+                    name += requireAs ? "As" : "";
+                    name += isWildcard ? "ResourceName" : typName;
+                    name += requirePlural ? "s" : "";
+                    return name;
+                }
             }
 
             /// <summary>The C# name of the string-typed property underlying this resource.</summary>
             public string UnderlyingPropertyName { get; }
 
-            /// <summary>The C# name of the resource property.</summary>
-            public string ResourcePropertyName { get; }
+            /// <summary>The C# name of the single-pattern resource property.</summary>
+            public string SingleResourcePropertyName { get; }
+
+            /// <summary>The C# name of the multi-pattern resource property.</summary>
+            public string MultiResourcePropertyName { get; }
 
             /// <summary>The resource definition for this field.</summary>
             public Definition ResourceDefinition { get; }
         }
 
-        public static IEnumerable<Definition> LoadResourceDefinitions(FileDescriptor fileDesc)
+        public static IReadOnlyList<Definition> LoadResourceDefinitionsByFileName(IEnumerable<FileDescriptor> descs)
         {
-            // TODO: Resource sets.
-            // Load file-level resource definitions.
-            var resourcesBySymbol1 = fileDesc.CustomOptions.TryGetRepeatedMessage<Resource>(ProtoConsts.FileOption.ResourceDefinition, out var resources) ?
-                resources.Select(resource => (symbol: MakeFullSymbol(fileDesc, null, resource.Symbol), resource)).ToList() :
-                Enumerable.Empty<(string symbol, Resource resource)>();
-            // Load message-level (shortcut) resource definitions.
-            var resourcesBySymbol2 = fileDesc.MessageTypes.SelectMany(msg =>
-                msg.Fields.InFieldNumberOrder().Select(field =>
-                    field.CustomOptions.TryGetMessage<Resource>(ProtoConsts.FieldOption.Resource, out var resource) ? resource : null)
-                    .Where(x => x != null)
-                    .Select(resource => (symbol: MakeFullSymbol(null, msg, resource.Symbol), resource)))
-                .ToList();
-            // Check no duplicates.
-            var duplicateSymbols = resourcesBySymbol1.Select(x => x.symbol).Concat(resourcesBySymbol2.Select(x => x.symbol))
-                .GroupBy(x => x).Where(x => x.Count() > 1).Select(x => x.Key).ToList();
-            if (duplicateSymbols.Any())
+            // Singles.
```

Comment:
Colon instead of period, for consistency with comments below?

Reply:
Done